### DISCUSSION TITLE
atomWithStorage issue since 2.3.0+

### DISCUSSION THREAD

**Discussion by fturmel** [pmndrs/jotai#2095] [Bug report] [answered]
I have an atom that interacts with an atomWithStorage that has stopped working correctly in 2.3.0+. There is a UI checkbox that toggles the `isDevState` atom but it is stuck at `false`. If I change the atomWithStorage's default value to `true`, the toggle works again.

Do I need to change my implementation, or is this a Jotai bug?

```ts
export const loggedInUserState = atom<Person | null>(null);

const _isDevState = atomWithStorage("_isDevState", false); // toggle works only if default is `true`

export const isDevState = atom(
  get => {
    const loggedInUser = get(loggedInUserState);
    if (!loggedInUser || loggedInUser.user?.isDev === false) return false;
    return get(_isDevState);
  },
  (get, set, value: boolean) => {
    set(_isDevState, value);
  }
);
```

**Comment by fturmel** [comment]
I think I've identified the edge case a bit better:
- the atomWithStorage needs to be `true` in local storage when the app starts
- the user isn't logged in right away, so the atom returns `false` initially
- after the user is logged in, `get(_isDevState)` returns `false` even though the atomWithStorage value is `true` in local storage (according to the Chrome dev tools Application/Local Storage panel).

**Comment by dai-shi** [comment]
Can you identify if #2029 changed the behavior?
Compare v2.2.3 and https://ci.codesandbox.io/status/pmndrs/jotai/pr/2029/builds/402968 (Find "Local Install Instructions").

**Comment by fturmel** [comment]
Thanks for the quick reply! 2.2.3 works properly, but PR 2029 is broken (same behavior I previously described since 2.3.0+).

**Reply by dai-shi** [reply]
Thanks for confirming. Hmm, there must be something I missed.
Would you be able to create a minimal reproduction with https://csb.jotai.org?
We then need to create a failing test case and tackle it.

**Reply by fturmel** [reply]
Sure, here you go: https://codesandbox.io/s/sharp-lamport-k9s8mp?file=/src/App.tsx

I've set the Jotai dep to 2.2.3 so you can see it working. Change it to 2.3.0 or 2.4.0 and you'll see the toggle input is not checked by default and stuck on that state.

**Reply by dai-shi** [reply]
Thanks. We probably need to dig it more.
I think we can reproduce the issue without atomWithStorage. Maybe, onMount is required.
I can try that later, but if you can try please feel free.



This is a workaround meanwhile: https://codesandbox.io/s/eager-bohr-c7g354?file=/src/App.tsx

**Reply by fturmel** [reply]
Thanks for the workaround, I can confirm that it works in the context of my app.

I tried and did not manage to reproduce the issue without atomWithStorage.

And to clarify, the `useEffect` in my previous reproduction was to make it more consistent. You can comment it out and it the bug is still reproducible as long as you reload the app and the localstorage value was last left to `true`.

Here's a screen recording showing the local storage inspector at the bottom and without the `useEffect`. If I load the app and the local storage value is `false`, the toggle works. If it's `true`, it gets stuck.

https://github.com/pmndrs/jotai/assets/4009209/3f2563ae-251e-486d-9207-f03be777bb2f

**Reply by dai-shi** [reply]
I would probably take the atomWithStorage code and remove the storage part and so on. Then the minimal repro will be without atomWithStorage.

**Reply by fturmel** [reply]
Oh, if I understand correctly I'm not familiar with the project's internals and won't be able to troubleshot at that level unfortunately.

I do have an additional piece of info that might be helpful. I found that the problem goes away if I set the `unstable_getOnInit` option to `true` on the atomWithStorage declaration like so:

```ts
const _isDevModeState = atomWithStorage("_isDevModeState", false, undefined, { unstable_getOnInit: true });
```

**Reply by dai-shi** [reply]
No worries. I can do that later, but things keep me busy at the moment. So, if anyone is interested, feel free to try.

I think it's a bug in #2029 and it's not specific to atomWithStorage.

**Reply by dai-shi** [reply]
Opened a tracking issue #2098.

**Reply by dai-shi** [reply]
#2099 will fix it.

**Reply by dai-shi** [reply]
thanks to @tmkx.

**Comment by dai-shi** [comment, marked as answer]
#2099 is merged.
